### PR TITLE
Refractor How Dialogue Is Handled Using DialogueManager

### LINKED ISSUE #206: InteractBox remains on NPC after leaving range and does not disappear

### PR DESCRIPTION
**Pull request by chadleytan**
Removed handling of playing dialogue in the interactable scripts and is now handled in `DialogueManager.cs`. Currently this can handle any conversation between characters as you simply just need to identify speakers with ID (e.g. can handle Birdie self dialogues or conversation between 2 NPCs).
Screen Shot 2020-03-15 at 9 09 31 PM
You would need to enter and fill out a `SpeakerContainer`:
* `Id` of the speaker (should be unique) which is a string
* Character Prefab
* NPC Voice Path (shouldn't cause an error if not filled)

Functionality to add any instantiated speakers/characters is supported but you just have to call `AddSpeaker()` in `DialogueManager.cs` which takes in a `SpeakerContainer`.

Other Main Changes:
* Removed `DialogueInteractable.cs`
* Removed `UITextOverlay` Scripts from character because functionality is now in the DialogueManage
* Can start a conversation calling `StartConversation()` in `DialogueManager.cs` which takes in a conversation
* Conversations work with ID now (ensure that Id exists in DialogueManager GameObject under systems
Screen Shot 2020-03-15 at 9 02 39 PM
* Added some public variables to `SpeakerUI` which needs to be updated in the scene
Screen Shot 2020-03-15 at 9 09 31 PM

Still yet to make the DialogueManager be able to handle multiple conversations but this should work for any functionality handling sequential conversations.

Fixes #206 
Related to #217

### CODE REVIEW

File: Assets/Prefabs/Characters/NurseNPC.prefab [hunk 1]
```diff
@@ -690,7 +673,7 @@ MonoBehaviour:
   m_Script: {fileID: 11500000, guid: a2a146b2f60f1492c836891f8a776db7, type: 3}
   m_Name: 
   m_EditorClassIdentifier: 
-  charTypeSpeed: 0.05
+  charTypeSpeed: -0.9
```
Comment (jurcaua) [inline Assets/Prefabs/Characters/NurseNPC.prefab]: negative typing speed??

File: Assets/Scripts/DialogueManager.cs [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+    // Update is called once per frame
+    void Update()
+    {
+        HandleInput();
+    }
+
+    private void HandleInput()
+    {
+        if (UIManager.Instance.IsGamePaused()) return;
```
Comment (jurcaua) [inline Assets/Scripts/DialogueManager.cs:70]: Make sure this still works:
* enter in range of an npc
* pause
* unpause using keyboard key Z
EXPECTED BEHAVIOUR: should not start conversation

had this bug before so we should make sure this still works properly

File: Assets/Scripts/DialogueManager.cs [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+        if (isTyping && Input.GetButtonDown(Constants.INPUT_INTERACTABLE_GETDOWN) && Time.frameCount != startFrame)
```
Comment (jurcaua) [inline Assets/Scripts/DialogueManager.cs:74]: make a comment about using `frameCount`? not immediately intuitive

File: Assets/Scripts/DialogueManager.cs [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+        {
+            if (!skipRequest)
+            {
+                skipRequest = true;
+            }
+            if (waitingForNext)
+            {
+                waitingForNext = false;
+            }
```
Comment (jurcaua) [inline Assets/Scripts/DialogueManager.cs]:
```suggestion
            skipRequest = true;
            waitingForNext = false;
```

File: Assets/Scripts/DialogueManager.cs [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+        if (!autoPlaying)
+        {
+            StartConversing();
+            GameManager.Instance.GetPlayerController().EnablePlayerInput = false;
+            AdvanceConversation();
+        }
+        else
+        {
+            coroutine = StartCoroutine(AutoplayConversation());
+        }
```
Comment (jurcaua) [inline Assets/Scripts/DialogueManager.cs]:
```suggestion
        if (convo.autoplayConversation)
        {
            TriggerAutoplay();
            // call this inside TriggerAutoplay
            // coroutine = StartCoroutine(AutoplayConversation());
        }
        else
        {
            StartConversing();
            GameManager.Instance.GetPlayerController().EnablePlayerInput = false;
            AdvanceConversation();
        }
```

File: Assets/Scripts/DialogueManager.cs [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+
+        return shouldShowLine;
+    }
+
+    void AdvanceConversation()
+    {
+        if (!ContinueConversation())
+        {
+            //Once the conversation is over 
+            //what happens
```
Comment (jurcaua) [inline Assets/Scripts/DialogueManager.cs]: I'd make this comment just span one line

File: Assets/Scripts/DialogueManager.cs [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+            StopCoroutine(currentTypingCoroutine);
+        }
+        currentTypingCoroutine = StartCoroutine(StartTypeText(text));
+
+        return currentTypingCoroutine;
+    }
+
+    private IEnumerator StartTypeText(string text)
+    {
+
```
Comment (jurcaua) [inline Assets/Scripts/DialogueManager.cs]: can we remove this empty hehe

File: Assets/Scripts/Interactables/Dialogue/Conversation.cs [hunk 8]
```diff
@@ -5,17 +5,17 @@
 [System.Serializable]
 public struct Line
 {
-    public bool isBirdie;
+    public string id;
 
     [TextArea(2, 5)]
```
Comment (jurcaua) [inline Assets/Scripts/Interactables/Dialogue/Conversation.cs]: Can we make it use this instead :)
https://github.com/dbrizov/NaughtyAttributes#resizabletextarea

(just put `using NaughtyAttributes;` at the top and you can use it)

Comment (jurcaua) [inline Assets/Scripts/Interactables/Dialogue/Conversation.cs]: Can we make it use this instead :)
https://github.com/dbrizov/NaughtyAttributes#resizabletextarea

(just put `using NaughtyAttributes;` at the top and you can use it)

File: Assets/Scripts/Interactables/Interactable.cs [hunk 9]
```diff
@@ -13,11 +13,11 @@ public class Interactable : MonoBehaviour, IInteractable
 
     protected bool enableInteract = true;
     protected bool interactableOn = false;
-    protected bool continueInteracting = false;
+    protected bool withinInteractRadius;
 
     public float interactRadius = Constants.INTERACT_POPUP_RADIUS;
 
-    protected bool withinInteractRadius;
+    //public bool withinInteractRadius;
```
Comment (jurcaua) [inline Assets/Scripts/Interactables/Interactable.cs]: can we remove this line since u have it above anyways?

**Review by jurcaua — CHANGES_REQUESTED** [review]
HOOOOOLY veryveryvery nice stuff im v impressed sir
thank u for your service 🙇 😻 

Made some comments!!
I'd also update the `---Systems---` prefab as well, since you probably added your DialogueManager prefab in there.

**Comment by masonachu** [comment]
awwwwyeeee making VO much easier to deal with thats my jam

File: Assets/Scripts/DialogueManager.cs [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+using System.Collections.Generic;
+using TMPro;
+using UnityEngine;
+
+[System.Serializable]
+public class SpeakerContainer
+{
+    public string id;
+    public GameObject speakerObject;
+    public string npcVoicePath;
```
Comment (masonachu) [inline Assets/Scripts/DialogueManager.cs:12]: if you put `[FMODUnity.EventRef]` above line 11, it'll allow me to use the FMOD event browser directly in the script so I won't need to go through the menus and search for the string. I can do that once this is merged but just so you know!

File: Assets/Prefabs/Characters/NurseNPC.prefab (hunk 1 above)
Comment (chadleytan) [reply Assets/Prefabs/Characters/NurseNPC.prefab]: loooool idk what happened there, but I'll fix

File: Assets/Scripts/DialogueManager.cs (hunk 2 above)
Comment (chadleytan) [reply Assets/Scripts/DialogueManager.cs:70]: Ok I just tested it out and it unpaused the screen without starting the convo

File: Assets/Scripts/DialogueManager.cs (hunk 10 above)
Comment (chadleytan) [reply Assets/Scripts/DialogueManager.cs:12]: Cool, I just added it!

File: Assets/Scripts/DialogueManager.cs (hunk 2 above)
Comment (chadleytan) [reply Assets/Scripts/DialogueManager.cs:70]: But I guess an issue now is making sure that the pause appears above the dialogue

**Review by jurcaua — APPROVED** [review]
cha ching!!! 🤑  great stuff man, good to merge

**Pull request merged**